### PR TITLE
task-259: add rootHash to logs

### PR DESCRIPTION
**Pull request by hypersolo**
## What does this MR do?

we needed a uniqueness on logs.
this is done in this pr.
i get the function and file name of log. then hash it and append it to log.params. 

## Check list  

- [ ] All Merge Request pipelines should be passed.
- [ ] Make sure all the changes are necessary and correct.
- [ ] Make sure no needed change missing (all necessary changes are made).
- [ ] Merge Request should contains only changes for one purpose (one feature or bug or refactoring).
- [ ] All necessary unit tests are written.
- [ ] Critical changes should be tested locally (or on staging).

### CODE REVIEW

File: src/utils/dexpressoLogger/logger.ts [hunk 1]
```diff
@@ -43,16 +59,24 @@ export default class Logger implements ILogger {
...
     data: { context: string; part?: string } & Record<string, string>,
   ): ILogger {
-    return this.engine.child({ data });
+    return new Logger({ engine: this.engine.child({ data }) });
   }
 }
 
+function getRootHash(): string {
+  const stack = new Error().stack ?? "";
+  const cleared = stack.split("\n")[3].replace(/[0-9]| /g, "");
```
Comment (WilliamHawkingStephen) [inline src/utils/dexpressoLogger/logger.ts:75]: - maybe a better name or comment that describe which line we select for more readability

File: src/utils/dexpressoLogger/logger.ts [hunk 2]
```diff
@@ -3,14 +3,27 @@ import { Logger as winstonLogger } from "winston";
...
+    lvl?: LogLevel;
+    engine?: winstonLogger;
+  }) {
+    const { service, lvl, engine } = data;
+    if (engine) {
+      this.engine = engine;
+    } else if (service && lvl) {
+      this.engine = createLoggerEngine(service, lvl);
+    } else {
+      throw new NotImplementedError("service or level is not provided");
```
Comment (WilliamHawkingStephen) [inline src/utils/dexpressoLogger/logger.ts:25]: - maybe we can add which service or level is cause error : 

`${service}, Level: ${lvl}`

File: src/utils/dexpressoLogger/logger.ts [hunk 3]
```diff
@@ -3,14 +3,27 @@ import { Logger as winstonLogger } from "winston";
...
   private static instances = new Map<string, ILogger>();
 
-  private constructor(service: string, lvl: LogLevel) {
-    this.engine = createLoggerEngine(service, lvl);
+  private constructor(data: {
+    service?: string;
+    lvl?: LogLevel;
+    engine?: winstonLogger;
+  }) {
+    const { service, lvl, engine } = data;
```
Comment (WilliamHawkingStephen) [inline src/utils/dexpressoLogger/logger.ts:19]: - suggestion: we can destructure the data in the function parameter : 
` private constructor({
  service,
  lvl,
  engine,
}: {
  service?: string;
  lvl?: LogLevel;
  engine?: winstonLogger;
}) {`

File: src/utils/dexpressoLogger/logger.ts (hunk 1 above)
Comment (hypersolo) [reply src/utils/dexpressoLogger/logger.ts]: you're right bro, https://github.com/dexpresso-protocol/core/pull/643/commits/f3615ef4878402fa6aee86892146f8824c817fd9

File: src/utils/dexpressoLogger/logger.ts (hunk 2 above)
Comment (hypersolo) [reply src/utils/dexpressoLogger/logger.ts:25]: it's called just in initialization process.
i just wanted to get sure that required props are provided.
i don't think logging undefined is a good idea!

**Review by WilliamHawkingStephen — APPROVED** [review]

File: src/utils/dexpressoLogger/logger.ts (hunk 3 above)
Comment (hypersolo) [reply src/utils/dexpressoLogger/logger.ts]: done, https://github.com/dexpresso-protocol/core/pull/643/commits/c43ef18981688f09ae0daa73027396ab1475b149

**Review by isaac-fibonacci — APPROVED** [review]

**Pull request merged**